pybind_interface/pybind_main.cpp

```diff
@@ -448,297 +448,199 @@ std::vector<std::complex<float>> qtrajectory_simulate(const py::dict &options) {
-  const float* ptr = input_vector.data();
-  auto f = [](unsigned n, unsigned m, uint64_t i, const float* ptr,
-              float* fsv) {
-    fsv[i] = ptr[i];
-  };
-  For(param.num_threads).Run(input_vector.size(), f, ptr, state.get());
-  state_space.NormalToInternalOrder(state);
+  bool is_noisy;
+  // Only one of these will be populated, as specified by is_noisy.
+  Circuit<Gate> circuit;
+  NoisyCircuit<Gate> ncircuit;
 
-  State scratch = StateSpace(1).Null();
-  std::vector<uint64_t> stat;
-  if (!Runner::Run(param, num_qubits, ncircuit, seed, scratch, state, stat)) {
-    IO::errorf(
-      "qtrajectory full state simulation of the circuit errored out.\n");
-    return StateSpace::Null();
-  }
-  return state;
-}
+  StateSpace state_space;
+  State state;
+  State scratch;
 
-py::array_t<float> state_to_python(
-    const py::dict &options, qsim::Simulator<For>::State&& state){
   unsigned num_qubits;
   unsigned num_threads;
-  try {
-    num_qubits = parseOptions<unsigned>(options, "n\0");
-    num_threads = parseOptions<unsigned>(options, "t\0");
-  } catch (const std::invalid_argument &exp) {
-    IO::errorf(exp.what());
-    return {};
-  }
-
-  qsim::Simulator<For>::StateSpace state_space(num_threads);
-  state_space.InternalToNormalOrder(state);
+  unsigned max_fused_size;
+  unsigned verbosity;
+  unsigned seed;
 
-  uint64_t fsv_size = 2 * (uint64_t{1} << num_qubits);
-  float* fsv = state.release();
-  auto capsule = py::capsule(
-      fsv, [](void *data) { delete reinterpret_cast<float *>(data); });
-  return py::array_t<float>(fsv_size, fsv, capsule);
-}
+  // Only set once initialization is complete.
+  bool is_valid;
+};
 
-// Externally-visible methods invoking the simulators above.
+// Methods for simulating full state vectors.
 
 py::array_t<float> qsim_simulate_fullstate(
     const py::dict &options, uint64_t input_state) {
-  return state_to_python(
-    options, _noiseless_sim_from_input_state(options, input_state));
+  return SimulatorHelper::simulate_fullstate(options, false, input_state);
 }
 
 py::array_t<float> qsim_simulate_fullstate(
     const py::dict &options, const py::array_t<float> &input_vector) {
-  return state_to_python(
-    options, _noiseless_sim_from_input_vector(options, input_vector));
+  return SimulatorHelper::simulate_fullstate(options, false, input_vector);
 }
 
 py::array_t<float> qtrajectory_simulate_fullstate(
     const py::dict &options, uint64_t input_state) {
-  return state_to_python(
-    options, _noisy_sim_from_input_state(options, input_state));
+  return SimulatorHelper::simulate_fullstate(options, true, input_state);
 }
 
 py::array_t<float> qtrajectory_simulate_fullstate(
     const py::dict &options, const py::array_t<float> &input_vector) {
-  return state_to_python(
-    options, _noisy_sim_from_input_vector(options, input_vector));
+  return SimulatorHelper::simulate_fullstate(options, true, input_vector);
 }
 
 // Methods for calculating expectation values.
```

Comment:
Perhaps you can add
```C++
// Methods for simulating amplitudes.
```
above and
```C++
// Methods for sampling.
```
and
```C++
// Method for running the hybrid simulator.
```
below or something like that.

Reply:
Done.